### PR TITLE
Add `return_offsets` and `truncate_input_tokens` tokenize API options

### PR DESCRIPTION
**Pull request by njhill**
By popular request

### CODE REVIEW

File: proto/generation.proto [hunk 1]
```diff
@@ -209,10 +213,17 @@ message TokenizeRequest {
...
 message TokenizeResponse {
+  message Offset {
+    uint32 start = 1;
+    uint32 end = 2;
+  }
+
   uint32 token_count = 1;
-  repeated string tokens = 2; // if include_tokens = true
 
-  // We'll possibly add more later
```
Comment (joerunde) [inline proto/generation.proto:215]: will we still possibly add more? 😉

File: router/src/grpc_server.rs [hunk 2]
```diff
@@ -339,17 +340,37 @@ impl GenerationService for GenerationServicer {
...
+        let truncate_to = match br.truncate_input_tokens {
+            0 => u32::MAX,
+            length => length,
+        };
+        let include_encoding = br.return_tokens || br.return_offsets;
+        let responses = try_join_all(br.requests.into_iter().map(|tr|
+            self.tokenizer.tokenize(tr.text, include_encoding).map_ok(
+                |(_, token_count, encoding)| {
+                    let token_count = token_count as u32;
+                    let from = token_count.saturating_sub(truncate_to) as usize;
```
Comment (joerunde) [inline router/src/grpc_server.rs:354]: TIL about `saturating_sub`, nice!

**Review by joerunde — APPROVED** [review]
looks pretty slick to me 😎

File: proto/generation.proto (hunk 1 above)
Comment (njhill) [reply proto/generation.proto:215]: never say never :)

**Pull request merged**